x509util/loginfo.go

```diff
@@ -0,0 +1,190 @@
+	return logInfoByKeyHash(ll, hc, NewLogInfoOverDNS)
+}
+
+func logInfoByKeyHash(ll *loglist.LogList, hc *http.Client, infoFactory func(*loglist.Log, *http.Client) (*LogInfo, error)) (map[[sha256.Size]byte]*LogInfo, error) {
+	result := make(map[[sha256.Size]byte]*LogInfo)
+	for _, log := range ll.Logs {
+		h := sha256.Sum256(log.Key)
+		li, err := infoFactory(&log, hc)
+		if err != nil {
+			return nil, err
+		}
+		result[h] = li
+	}
+	return result, nil
+}
+
+// merkleVerifier is used to verify Merkle tree calculations.
+var merkleVerifier = merkletree.NewMerkleVerifier(func(data []byte) []byte {
+	hash := sha256.Sum256(data)
+	return hash[:]
+})
+
+// LastSTH returns the last STH known for the log.
+func (li *LogInfo) LastSTH() *ct.SignedTreeHead {
+	li.mu.RLock()
+	defer li.mu.RUnlock()
+	return li.lastSTH
+}
+
+// SetSTH sets the last STH known for the log.
+func (li *LogInfo) SetSTH(sth *ct.SignedTreeHead) {
+	li.mu.Lock()
+	defer li.mu.Unlock()
+	li.lastSTH = sth
+}
+
+// VerifySCTSignature checks the signature in the SCT matches the given leaf (adjusted for the
+// timestamp in the SCT) and log.
+func (li *LogInfo) VerifySCTSignature(sct ct.SignedCertificateTimestamp, leaf ct.MerkleTreeLeaf) error {
+	leaf.TimestampedEntry.Timestamp = sct.Timestamp
+	if err := li.Verifier.VerifySCTSignature(sct, ct.LogEntry{Leaf: leaf}); err != nil {
+		return fmt.Errorf("failed to verify SCT signature from log %q: %v", li.Description, err)
+	}
+	return nil
+}
+
+// VerifyInclusionLatest checks that the given Merkle tree leaf, adjusted for the provided timestamp,
+// is present in the latest known tree size of the log.  If no tree size for the log is known, it will
+// be queried.
+func (li *LogInfo) VerifyInclusionLatest(ctx context.Context, leaf ct.MerkleTreeLeaf, timestamp uint64) (int64, error) {
+	sth := li.LastSTH()
+	if sth == nil {
+		var err error
+		sth, err = li.Client.GetSTH(ctx)
+		if err != nil {
+			return -1, fmt.Errorf("failed to get current STH for %q log: %v", li.Description, err)
+		}
+		li.SetSTH(sth)
+	}
+	return li.VerifyInclusionAt(ctx, leaf, timestamp, sth.TreeSize, sth.SHA256RootHash[:])
+}
+
+// VerifyInclusion checks that the given Merkle tree leaf, adjusted for the provided timestamp,
+// is present in the current tree size of the log.
+func (li *LogInfo) VerifyInclusion(ctx context.Context, leaf ct.MerkleTreeLeaf, timestamp uint64) (int64, error) {
+	sth, err := li.Client.GetSTH(ctx)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get current STH for %q log: %v", li.Description, err)
+	}
+	li.SetSTH(sth)
+	return li.VerifyInclusionAt(ctx, leaf, timestamp, sth.TreeSize, sth.SHA256RootHash[:])
+}
+
+// VerifyInclusionAt checks that the given Merkle tree leaf, adjusted for the provided timestamp,
+// is present in the given tree size & root hash of the log.
+func (li *LogInfo) VerifyInclusionAt(ctx context.Context, leaf ct.MerkleTreeLeaf, timestamp, treeSize uint64, rootHash []byte) (int64, error) {
```

Comment:
Could you add to the doc for this function (and the other verify inclusion ones) to say what the int64 that is returned represents? (Looks like it's the index returned alongside the proof by get-proof-by-hash?)

Reply:
Done.